### PR TITLE
KVM: Propagating changes on host parameters to the agents

### PR DESCRIPTION
**Pull request by ustcweizhou**
## Description

When we change a global setting and it needs to be applied to all kvm agents, it would be nice that the new value can be propagated to all kvm agents (and save in agent.properties if needed) so we do not need to manually restart cloudstack agent or change agent.properties.

it supports only "router.aggregation.command.each.timeout" for now.

In the future, it would be good to add a new column in "configurations" table to determine whether the configuration needs to be propagated to kvm agents.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?

(1) change router.aggregation.command.each.timeout in global setting
(2) check agent.properties to see if the value is changed as well.

### CODE REVIEW

**Comment by svenvogel** [comment]
@ustcweizhou small question. do you mean with "configuration" ... global settings? how can we add more options which already exists in the agent.properties with your code?

**Comment by ustcweizhou** [comment]
@svenvogel 
if global configuration already exists, then add it to line 1859
https://github.com/apache/cloudstack/pull/3491/files#diff-47d8b846590fa4ae09924a4e9569328bR1859

If global configuration does not exist, we need to add a ConfigKey at first.
similar as https://github.com/apache/cloudstack/blob/4.11/engine/orchestration/src/com/cloud/agent/manager/AgentManagerImpl.java#L193-L210

**Comment by wido** [comment]
Thanks! This is very welcome. I also looked into this for #3500 , but my problem with it is that it's very hardcoded. Both on the Mgmt server and Agent side.

The name for the keys are repeated multiple times and that can cause problems. It doesn't really scale.

How could we handle that?

**Comment by svenvogel** [comment]
@wido @ustcweizhou yes could be cool if we find a method which is lesser hardcoded. maybe a global setting "," comma separated?

**Comment by rhtyd** [comment]
@ustcweizhou I tried changing the base branch to master and 4.13, there are conflicts. 4.11 is very old, can you rebase to master?

**Comment by rhtyd** [comment]
@ustcweizhou can you rebase to 4.13? I've tentatively set the milestone to 4.13.1.0.

**Comment by weizhouapache** [comment]
rebased with latest master

**Comment by DaanHoogland** [comment]
@weizhouapache do we want this on 4.13 or on master. it is now based on master but targetted towards 4.13. Can you rebase or should we change the target?

**Comment by weizhouapache** [comment]
@DaanHoogland a, my bad. I will rebase with 4.13

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@nvazquez can you review please. I see one ha error biut no relation to the code, thanks.

**Comment by nvazquez** [comment]
Hi @weizhouapache, some of this functionality is already existing for the software load balancing PR, in which changing the value of 'host', 'indirect.agent.lb.algorithm' or 'indirect.agent.lb.algorithm' triggers a command being sent to each connected agent. Can you please check this existing functionality to reuse some parts of it perhaps?

**Comment by weizhouapache** [comment]
> Hi @weizhouapache, some of this functionality is already existing for the software load balancing PR, in which changing the value of 'host', 'indirect.agent.lb.algorithm' or 'indirect.agent.lb.algorithm' triggers a command being sent to each connected agent. Can you please check this existing functionality to reuse some parts of it perhaps?

@nvazquez yes, there are some similar codes in server/src/main/java/org/apache/cloudstack/agent/lb/IndirectAgentLBServiceImpl.java

IMHO, It make more sense to add these code in AgentManagerImpl.java instead of IndirectAgentLBServiceImpl.java. What do you think ?
I can remove some duplicated code in IndirectAgentLBServiceImpl.java.

**Comment by rhtyd** [comment]
@weizhouapache @nvazquez a refactoring would be great to consolidate related/common code.

**Comment by weizhouapache** [comment]
@nvazquez @rhtyd 
I have moved the method to initialize messagebus listener from AgentManagerImpl.java  and IndirectAgentLBServiceImpl to ConfigurationManagerImpl.java 

tested
(1) change global setting router.aggregation.command.each.timeout, the value is sent to all cloudstack-agent and saved in agent.properties
(2) change global setting host or indirect.agent.lb.algorithm, the change is sent to cloudstack-agent, agent.properties is changed as well.

**Comment by weizhouapache** [comment]
unit tests failed. I will fix it

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by andrijapanicsb** [comment]
@weizhouapache and others, I don't want to be picky, but this is not really a bug fix, should we target it against master only?

**Comment by weizhouapache** [comment]
> @weizhouapache and others, I don't want to be picky, but this is not really a bug fix, should we target it against master only?

@andrijapanicsb agree with merging to master instead of 4.13.
I will change dest branch to master if nobody objects @DaanHoogland @rhtyd

**Comment by rhtyd** [comment]
Yes master/4.14 is good.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@wido, @svenvogel did you review/test?

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@nvazquez @DaanHoogland @andrijapanicsb can you review?

**Comment by DaanHoogland** [comment]
tested according to Wei's spec and looks good. On advise from @nvazquez I also tested if the other pushed properties where still fine after the push:
`host=10.2.2.171,10.2.2.168@static` 👍

File: engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java [hunk 1]
```diff
@@ -1788,4 +1788,45 @@ public int getTimeout() {
...
+            }
+        }
+    }
+
+    @Override
+    public void propagateChangeToAgents() {
+        s_logger.debug("Propagating changes on host parameters to the agents");
+        Map<Long, List<Long>> hostsPerZone = getHostsPerZone();
+        Map<String, String> params = new HashMap<String, String>();
+        params.put("router.aggregation.command.each.timeout", _configDao.getValue("router.aggregation.command.each.timeout"));
```
Comment (nvazquez) [inline engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java]: What about removing this one from here and adding `params` as parameter?

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 2]
```diff
@@ -461,6 +471,24 @@ private void overProvisioningFactorsForValidation() {
...
+        messageBus.subscribe(EventTypes.EVENT_CONFIGURATION_VALUE_EDIT, new MessageSubscriber() {
+            @Override
+            public void onPublishMessage(String serderAddress, String subject, Object args) {
+                String globalSettingUpdated = (String) args;
+                if (Strings.isNullOrEmpty(globalSettingUpdated)) {
+                    return;
+                }
+                if (globalSettingUpdated.equals(ApiServiceConfiguration.ManagementServerAddresses.key()) ||
+                        globalSettingUpdated.equals(IndirectAgentLBServiceImpl.IndirectAgentLBAlgorithm.key())) {
+                    _indirectAgentLB.propagateMSListToAgents();
```
Comment (nvazquez) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java]: I think we can unify both calls as they do basically the same

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 3]
```diff
@@ -461,6 +471,24 @@ private void overProvisioningFactorsForValidation() {
...
+                } else if (globalSettingUpdated.equals(Config.RouterAggregationCommandEachTimeout.toString())) {
+                    _agentManager.propagateChangeToAgents();
```
Comment (nvazquez) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java]: ^^

**Comment by DaanHoogland** [comment]
merging this as is but created #3896 for follow up work

File: engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java (hunk 1 above)
Comment (weizhouapache) [reply engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java]: @nvazquez it is absolutely good.
I will create another PR. this PR has already been merged into 4.13 and 4.14 ...

**Pull request merged**